pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,882 @@
+		}
+		return nil, nil
+	}
+	if klog.V(5).Enabled() {
+		var sample []string
+		for i := 0; i < 10 && i < len(potentialNodes); i++ {
+			sample = append(sample, potentialNodes[i].Node().Name)
+		}
+		klog.Infof("%v potential nodes for preemption, first %v are: %v", len(potentialNodes), len(sample), sample)
+	}
+
+	pdbs, err := getPodDisruptionBudgets(pdbLister)
+	if err != nil {
+		return nil, err
+	}
+	return dryRunPreemption(ctx, ph, state, pod, potentialNodes, pdbs), nil
+}
+
+// dryRunPreemption simulates Preemption logic on <potentialNodes> in parallel,
+// and returns all possible preemption candidates.
+func dryRunPreemption(ctx context.Context, fh framework.PreemptHandle, state *framework.CycleState,
+	pod *v1.Pod, potentialNodes []*framework.NodeInfo, pdbs []*policy.PodDisruptionBudget) []defaultpreemption.Candidate {
+	var resultLock sync.Mutex
+	var candidates []defaultpreemption.Candidate
+	snapshotState, err := getSnapshotState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", SnapshotStateKey, err)
+		return nil
+	}
+	preFilterState, err := getPreFilterState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", preFilterStateKey, err)
+		return nil
+	}
+
+	checkNode := func(i int) {
+		nodeInfoCopy := potentialNodes[i].Clone()
+		stateCopy := state.Clone()
+		pods, numPDBViolations, fits := selectVictimsOnNode(ctx, fh, stateCopy, pod, nodeInfoCopy, pdbs, snapshotState, preFilterState)
+		if fits {
+			resultLock.Lock()
+			victims := extenderv1.Victims{
+				Pods:             pods,
+				NumPDBViolations: int64(numPDBViolations),
+			}
+			c := candidate{
+				victims: &victims,
+				name:    nodeInfoCopy.Node().Name,
+			}
+			candidates = append(candidates, &c)
+			resultLock.Unlock()
+		}
+	}
+	cutil.Until(ctx, len(potentialNodes), checkNode)
+	return candidates
+}
+
+// nodesWherePreemptionMightHelp returns a list of nodes with failed predicates
+// that may be satisfied by removing pods from the node.
+func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) []*framework.NodeInfo {
+	var potentialNodes []*framework.NodeInfo
+	for _, node := range nodes {
+		name := node.Node().Name
+		// We reply on the status by each plugin - 'Unschedulable' or 'UnschedulableAndUnresolvable'
+		// to determine whether preemption may help or not on the node.
+		if m[name].Code() == framework.UnschedulableAndUnresolvable {
+			continue
+		}
+		potentialNodes = append(potentialNodes, node)
+	}
+	return potentialNodes
+}
+
+func selectVictimsOnNode(
+	ctx context.Context,
+	ph framework.PreemptHandle,
+	state *framework.CycleState,
+	pod *v1.Pod,
+	nodeInfo *framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget,
+	snapshotState *SnapshotState,
+	preFilterState *PreFilterState,
+) ([]*v1.Pod, int, bool) {
+	var potentialVictims []*v1.Pod
+	var moreThanMinWithPreemptor bool
+
+	removePod := func(rp *v1.Pod) error {
```

Comment:
As `PreFilterExtensions` is not implemented, removePod and addPod won't change the PreFilterState in terms of ElasticQuota. So does that mean upon a PreFilter error raised by this plugin, the dry run would always fail?